### ISSUE TITLE
no implicit conversion of Sprockets::Asset into String

### ISSUE THREAD

**Issue by mosinski** [franzheidl/sass-inline-svg#6]
I just setup gem on my app
and try to use it in my example-file.scss like this:
`background-image: inline_svg('swimming.svg');`

and I get error: `no implicit conversion of Sprockets::Asset into String`

Normally successfully I use:
`background-image: image-url('cycling.svg');`

Is this gem does not work anymore?

**Comment by franzheidl** [comment]
What version of the gem are you using? What environment (Rails? What version?)

**Comment by mosinski** [comment]
Ruby **2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]**
Rails **4.2.6**
Sass-rails **5.0.4**
Sass_inline_svg **0.0.4**
Sprockets **3.6.0**
Sprockets-rails **3.0.4**
Environment: **development**

Full Trace

```
sass_inline_svg (0.0.4) lib/sass_inline_svg.rb:47:in `readable?'
sass_inline_svg (0.0.4) lib/sass_inline_svg.rb:47:in `_readFile'
sass_inline_svg (0.0.4) lib/sass_inline_svg.rb:24:in `inline_svg'
sass (3.4.22) lib/sass/script/tree/funcall.rb:143:in `_perform'
sass (3.4.22) lib/sass/script/tree/node.rb:58:in `perform'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:397:in `visit_prop'
sass (3.4.22) lib/sass/tree/visitors/base.rb:36:in `visit'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:160:in `block in visit'
sass (3.4.22) lib/sass/stack.rb:79:in `block in with_base'
sass (3.4.22) lib/sass/stack.rb:115:in `with_frame'
sass (3.4.22) lib/sass/stack.rb:79:in `with_base'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:160:in `visit'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:433:in `block (2 levels) in visit_rule'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:433:in `map'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:433:in `block in visit_rule'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:181:in `with_environment'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:431:in `visit_rule'
sass (3.4.22) lib/sass/tree/visitors/base.rb:36:in `visit'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:160:in `block in visit'
sass (3.4.22) lib/sass/stack.rb:79:in `block in with_base'
sass (3.4.22) lib/sass/stack.rb:115:in `with_frame'
sass (3.4.22) lib/sass/stack.rb:79:in `with_base'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:160:in `visit'
sass (3.4.22) lib/sass/tree/visitors/base.rb:52:in `block in visit_children'
sass (3.4.22) lib/sass/tree/visitors/base.rb:52:in `map'
sass (3.4.22) lib/sass/tree/visitors/base.rb:52:in `visit_children'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:169:in `block in visit_children'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:181:in `with_environment'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:168:in `visit_children'
sass (3.4.22) lib/sass/tree/visitors/base.rb:36:in `block in visit'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:188:in `visit_root'
sass (3.4.22) lib/sass/tree/visitors/base.rb:36:in `visit'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:159:in `visit'
sass (3.4.22) lib/sass/tree/visitors/perform.rb:8:in `visit'
sass (3.4.22) lib/sass/tree/root_node.rb:36:in `css_tree'
sass (3.4.22) lib/sass/tree/root_node.rb:20:in `render'
sass (3.4.22) lib/sass/engine.rb:281:in `render'
sass-rails (5.0.4) lib/sass/rails/template.rb:47:in `evaluate'
tilt (2.0.2) lib/tilt/template.rb:99:in `render'
sprockets (3.6.0) lib/sprockets/legacy_tilt_processor.rb:25:in `call'
sprockets (3.6.0) lib/sprockets/processor_utils.rb:75:in `call_processor'
sprockets (3.6.0) lib/sprockets/processor_utils.rb:57:in `block in call_processors'
sprockets (3.6.0) lib/sprockets/processor_utils.rb:56:in `reverse_each'
sprockets (3.6.0) lib/sprockets/processor_utils.rb:56:in `call_processors'
sprockets (3.6.0) lib/sprockets/loader.rb:134:in `load_from_unloaded'
sprockets (3.6.0) lib/sprockets/loader.rb:60:in `block in load'
sprockets (3.6.0) lib/sprockets/loader.rb:317:in `fetch_asset_from_dependency_cache'
sprockets (3.6.0) lib/sprockets/loader.rb:44:in `load'
sprockets (3.6.0) lib/sprockets/cached_environment.rb:20:in `block in initialize'
sprockets (3.6.0) lib/sprockets/cached_environment.rb:47:in `load'
sprockets (3.6.0) lib/sprockets/bundle.rb:23:in `block in call'
sprockets (3.6.0) lib/sprockets/utils.rb:183:in `dfs'
sprockets (3.6.0) lib/sprockets/bundle.rb:24:in `call'
sprockets (3.6.0) lib/sprockets/processor_utils.rb:75:in `call_processor'
sprockets (3.6.0) lib/sprockets/processor_utils.rb:57:in `block in call_processors'
sprockets (3.6.0) lib/sprockets/processor_utils.rb:56:in `reverse_each'
sprockets (3.6.0) lib/sprockets/processor_utils.rb:56:in `call_processors'
sprockets (3.6.0) lib/sprockets/loader.rb:134:in `load_from_unloaded'
sprockets (3.6.0) lib/sprockets/loader.rb:60:in `block in load'
sprockets (3.6.0) lib/sprockets/loader.rb:317:in `fetch_asset_from_dependency_cache'
sprockets (3.6.0) lib/sprockets/loader.rb:44:in `load'
sprockets (3.6.0) lib/sprockets/cached_environment.rb:20:in `block in initialize'
sprockets (3.6.0) lib/sprockets/cached_environment.rb:47:in `load'
sprockets (3.6.0) lib/sprockets/base.rb:66:in `find_asset'
sprockets (3.6.0) lib/sprockets/base.rb:92:in `[]'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:341:in `find_asset'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:333:in `find_debug_asset'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:216:in `block in lookup_debug_asset'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:229:in `block in resolve_asset'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:228:in `each'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:228:in `detect'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:228:in `resolve_asset'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:215:in `lookup_debug_asset'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:157:in `block in stylesheet_link_tag'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:156:in `map'
sprockets-rails (3.0.4) lib/sprockets/rails/helper.rb:156:in `stylesheet_link_tag'
app/views/layouts/application.html.erb:7:in `_app_views_layouts_application_html_erb___1833742688027284267_70267780257940'
actionview (4.2.6) lib/action_view/template.rb:145:in `block in render'
activesupport (4.2.6) lib/active_support/notifications.rb:166:in `instrument'
actionview (4.2.6) lib/action_view/template.rb:333:in `instrument'
actionview (4.2.6) lib/action_view/template.rb:143:in `render'
actionview (4.2.6) lib/action_view/renderer/template_renderer.rb:66:in `render_with_layout'
actionview (4.2.6) lib/action_view/renderer/template_renderer.rb:52:in `render_template'
actionview (4.2.6) lib/action_view/renderer/template_renderer.rb:14:in `render'
actionview (4.2.6) lib/action_view/renderer/renderer.rb:46:in `render_template'
actionview (4.2.6) lib/action_view/renderer/renderer.rb:27:in `render'
actionview (4.2.6) lib/action_view/rendering.rb:100:in `_render_template'
actionpack (4.2.6) lib/action_controller/metal/streaming.rb:217:in `_render_template'
actionview (4.2.6) lib/action_view/rendering.rb:83:in `render_to_body'
actionpack (4.2.6) lib/action_controller/metal/rendering.rb:32:in `render_to_body'
actionpack (4.2.6) lib/action_controller/metal/renderers.rb:37:in `render_to_body'
actionpack (4.2.6) lib/abstract_controller/rendering.rb:25:in `render'
actionpack (4.2.6) lib/action_controller/metal/rendering.rb:16:in `render'
actionpack (4.2.6) lib/action_controller/metal/instrumentation.rb:44:in `block (2 levels) in render'
activesupport (4.2.6) lib/active_support/core_ext/benchmark.rb:12:in `block in ms'
/usr/local/rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/benchmark.rb:308:in `realtime'
activesupport (4.2.6) lib/active_support/core_ext/benchmark.rb:12:in `ms'
actionpack (4.2.6) lib/action_controller/metal/instrumentation.rb:44:in `block in render'
actionpack (4.2.6) lib/action_controller/metal/instrumentation.rb:87:in `cleanup_view_runtime'
activerecord (4.2.6) lib/active_record/railties/controller_runtime.rb:25:in `cleanup_view_runtime'
actionpack (4.2.6) lib/action_controller/metal/instrumentation.rb:43:in `render'
remotipart (1.2.1) lib/remotipart/render_overrides.rb:14:in `render_with_remotipart'
responders (2.2.0) lib/action_controller/responder.rb:238:in `default_render'
responders (2.2.0) lib/action_controller/responder.rb:170:in `to_html'
responders (2.2.0) lib/action_controller/responder.rb:163:in `respond'
responders (2.2.0) lib/action_controller/responder.rb:156:in `call'
responders (2.2.0) lib/action_controller/respond_with.rb:205:in `respond_with'
devise (4.1.0) app/controllers/devise/sessions_controller.rb:12:in `new'
actionpack (4.2.6) lib/action_controller/metal/implicit_render.rb:4:in `send_action'
actionpack (4.2.6) lib/abstract_controller/base.rb:198:in `process_action'
actionpack (4.2.6) lib/action_controller/metal/rendering.rb:10:in `process_action'
actionpack (4.2.6) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
activesupport (4.2.6) lib/active_support/callbacks.rb:117:in `call'
activesupport (4.2.6) lib/active_support/callbacks.rb:555:in `block (2 levels) in compile'
activesupport (4.2.6) lib/active_support/callbacks.rb:505:in `call'
activesupport (4.2.6) lib/active_support/callbacks.rb:92:in `__run_callbacks__'
activesupport (4.2.6) lib/active_support/callbacks.rb:778:in `_run_process_action_callbacks'
activesupport (4.2.6) lib/active_support/callbacks.rb:81:in `run_callbacks'
actionpack (4.2.6) lib/abstract_controller/callbacks.rb:19:in `process_action'
actionpack (4.2.6) lib/action_controller/metal/rescue.rb:29:in `process_action'
actionpack (4.2.6) lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
activesupport (4.2.6) lib/active_support/notifications.rb:164:in `block in instrument'
activesupport (4.2.6) lib/active_support/notifications/instrumenter.rb:20:in `instrument'
activesupport (4.2.6) lib/active_support/notifications.rb:164:in `instrument'
actionpack (4.2.6) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
actionpack (4.2.6) lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
activerecord (4.2.6) lib/active_record/railties/controller_runtime.rb:18:in `process_action'
actionpack (4.2.6) lib/abstract_controller/base.rb:137:in `process'
actionview (4.2.6) lib/action_view/rendering.rb:30:in `process'
actionpack (4.2.6) lib/action_controller/metal.rb:196:in `dispatch'
actionpack (4.2.6) lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
actionpack (4.2.6) lib/action_controller/metal.rb:237:in `block in action'
actionpack (4.2.6) lib/action_dispatch/routing/route_set.rb:74:in `dispatch'
actionpack (4.2.6) lib/action_dispatch/routing/route_set.rb:43:in `serve'
actionpack (4.2.6) lib/action_dispatch/routing/mapper.rb:49:in `serve'
actionpack (4.2.6) lib/action_dispatch/journey/router.rb:43:in `block in serve'
actionpack (4.2.6) lib/action_dispatch/journey/router.rb:30:in `each'
actionpack (4.2.6) lib/action_dispatch/journey/router.rb:30:in `serve'
actionpack (4.2.6) lib/action_dispatch/routing/route_set.rb:817:in `call'
warden (1.2.6) lib/warden/manager.rb:35:in `block in call'
warden (1.2.6) lib/warden/manager.rb:34:in `catch'
warden (1.2.6) lib/warden/manager.rb:34:in `call'
rack (1.6.4) lib/rack/etag.rb:24:in `call'
rack (1.6.4) lib/rack/conditionalget.rb:25:in `call'
rack (1.6.4) lib/rack/head.rb:13:in `call'
remotipart (1.2.1) lib/remotipart/middleware.rb:27:in `call'
actionpack (4.2.6) lib/action_dispatch/middleware/params_parser.rb:27:in `call'
actionpack (4.2.6) lib/action_dispatch/middleware/flash.rb:260:in `call'
rack (1.6.4) lib/rack/session/abstract/id.rb:225:in `context'
rack (1.6.4) lib/rack/session/abstract/id.rb:220:in `call'
actionpack (4.2.6) lib/action_dispatch/middleware/cookies.rb:560:in `call'
activerecord (4.2.6) lib/active_record/query_cache.rb:36:in `call'
activerecord (4.2.6) lib/active_record/connection_adapters/abstract/connection_pool.rb:653:in `call'
activerecord (4.2.6) lib/active_record/migration.rb:377:in `call'
actionpack (4.2.6) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
activesupport (4.2.6) lib/active_support/callbacks.rb:88:in `__run_callbacks__'
activesupport (4.2.6) lib/active_support/callbacks.rb:778:in `_run_call_callbacks'
activesupport (4.2.6) lib/active_support/callbacks.rb:81:in `run_callbacks'
actionpack (4.2.6) lib/action_dispatch/middleware/callbacks.rb:27:in `call'
actionpack (4.2.6) lib/action_dispatch/middleware/reloader.rb:73:in `call'
actionpack (4.2.6) lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
better_errors (2.1.1) lib/better_errors/middleware.rb:59:in `call'
actionpack (4.2.6) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
web-console (2.3.0) lib/web_console/middleware.rb:20:in `block in call'
web-console (2.3.0) lib/web_console/middleware.rb:18:in `catch'
web-console (2.3.0) lib/web_console/middleware.rb:18:in `call'
actionpack (4.2.6) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
railties (4.2.6) lib/rails/rack/logger.rb:38:in `call_app'
railties (4.2.6) lib/rails/rack/logger.rb:20:in `block in call'
activesupport (4.2.6) lib/active_support/tagged_logging.rb:68:in `block in tagged'
activesupport (4.2.6) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (4.2.6) lib/active_support/tagged_logging.rb:68:in `tagged'
railties (4.2.6) lib/rails/rack/logger.rb:20:in `call'
quiet_assets (1.1.0) lib/quiet_assets.rb:27:in `call_with_quiet_assets'
actionpack (4.2.6) lib/action_dispatch/middleware/request_id.rb:21:in `call'
rack (1.6.4) lib/rack/methodoverride.rb:22:in `call'
rack (1.6.4) lib/rack/runtime.rb:18:in `call'
activesupport (4.2.6) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
rack (1.6.4) lib/rack/lock.rb:17:in `call'
actionpack (4.2.6) lib/action_dispatch/middleware/static.rb:120:in `call'
rack (1.6.4) lib/rack/sendfile.rb:113:in `call'
railties (4.2.6) lib/rails/engine.rb:518:in `call'
railties (4.2.6) lib/rails/application.rb:165:in `call'
rack (1.6.4) lib/rack/content_length.rb:15:in `call'
puma (2.11.1) lib/puma/server.rb:507:in `handle_request'
puma (2.11.1) lib/puma/server.rb:375:in `process_client'
puma (2.11.1) lib/puma/server.rb:262:in `block in run'
puma (2.11.1) lib/puma/thread_pool.rb:104:in `block in spawn_thread'
```

**Comment by franzheidl** [comment]
Hmm, weird. The file exists and is a readable (text-) file, not a binary?
I'll try to reproduce / investigate.

**Comment by mosinski** [comment]
do you think that the fault of a file?
I use this one: http://imgh.us/swimming.svg
but I tried another, unfortunately with the same result: http://imgh.us/sample_5.svg

**Comment by franzheidl** [comment]
Just wanted to make sure this is a real issue and unrelated to the file. Appears it is a real issue.

**Comment by mosinski** [comment]
Hi, any progress in resolving error :grey_question:
cuz I really want to use this gem :)

**Comment by franzheidl** [comment]
Not yet, very busy at the mo, hoping to get to it today or tomorrow.

**Comment by franzheidl** [comment]
Hi, I'm having trouble (and not the time) to reproduce your exact environment.

To make this quicker for you, could you please check if this fix is working by using the gem from a git branch in your Gemfile:

`gem 'sass_inline_svg', :git => 'https://github.com/franzheidl/sass-inline-svg.git', :branch => 'fix-no-implicit-conversion'`

Let me know.

**Comment by mosinski** [comment]
I'm suprised because the error does not show :+1: awesome :smiley: 
but when I trying to used fill option:

`background-image: inline_svg('tournament.svg', ( fillcolor: 'red' ));`

It generates me:

``` html
data:image/svg+xml;charset=utf-8,<svg xmlns%3D"http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg" xmlns%3Axlink%3D"http%3A%2F%2Fwww.w3.org%2F1999%2Fxlink" version%3D"1.1" x%3D"0px" y%3D"0px" viewBox%3D"0 0 100 100" enable-background%3D"new 0 0 100 100" xml%3Aspace%3D"preserve"><path d%3D"M5.468%2C97.271c0%2C1.506-1.223%2C2.733-2.734%2C2.733l0%2C0C1.223%2C100.004%2C0%2C98.776%2C0%2C97.271V2.162c0-1.51%2C1.223-2.734%2C2.734-2.734  l0%2C0c1.511%2C0%2C2.734%2C1.224%2C2.734%2C2.734V97.271z"><%2Fpath><rect x%3D"5.468" y%3D"3.925" width%3D"11.717" height%3D"11.716"><%2Frect><rect x%3D"17.156" y%3D"15.612" width%3D"11.714" height%3D"11.717"><%2Frect><rect x%3D"5.468" y%3D"27.297" width%3D"11.717" height%3D"11.717"><%2Frect><rect x%3D"17.156" y%3D"38.986" width%3D"11.714" height%3D"11.716"><%2Frect><rect x%3D"28.87" y%3D"3.925" width%3D"11.714" height%3D"11.716"><%2Frect><rect x%3D"40.563" y%3D"15.612" width%3D"11.709" height%3D"11.717"><%2Frect><rect x%3D"28.87" y%3D"27.297" width%3D"11.714" height%3D"11.717"><%2Frect><rect x%3D"40.563" y%3D"38.986" width%3D"11.709" height%3D"11.716"><%2Frect><rect x%3D"52.272" y%3D"3.925" width%3D"11.721" height%3D"11.716"><%2Frect><rect x%3D"63.966" y%3D"15.612" width%3D"11.713" height%3D"11.717"><%2Frect><rect x%3D"52.272" y%3D"27.297" width%3D"11.721" height%3D"11.717"><%2Frect><rect x%3D"63.966" y%3D"38.986" width%3D"11.713" height%3D"11.716"><%2Frect><path d%3D"M78.327%2C53.34H1.408V1.529h76.919V53.34z M4.06%2C50.691h71.619V4.18H4.06V50.691z"><%2Fpath><%2Fsvg>
```

and it's still black, I checked on ff and chrome disame result.. the file is wrong ? I want to change black areas to anather color

**Comment by franzheidl** [comment]
**Good to hear the fix works, will release a new version of the gem soon.**

Re what you call the 'fill-option': This does not necessarily change the fill-color, what it does is string replacement (which you can use to change the fill-color or everything else you want).
The key of the replacement object is the string you want to replace in the SVG, the value is the string to replace it with, so that

`background-image: inline_svg('tournament.svg', ( fillcolor: 'red' ));`

changes

`<rect fill="fillcolor" />`

to

`<rect fill="red" />`

Think like using **variables in your SVGs**. 

Or **replace existing strings** if you're sure you want to replace all occurences in the SVG:

`background-image: inline_svg('tournament.svg', ( '#000000': 'red' ));`

changes

`<rect fill="#000000" />`

to

`<rect fill="red" />`

Your example SVG does not have `fill` defined anywhere, so a) it defaults to black and b) nothing is replaced.

**Comment by mosinski** [comment]
ooh now I understand thanks for the explanation :)
I will improved the file, and **really thanks for Your time**

**Comment by franzheidl** [comment]
No worries, happy it works for you again.  👍🏼

**Comment by smellyshovel** [comment]
Hi. Why didn't you release a new version yet? This gem is just what I (and I think not only I) needed. But now I need to use 0.0.4 version, which produces the same problem for me.
I wait the new version and hope on you. Thanks.

**Comment by smellyshovel** [comment]
Oh, sorry, I didn't notice that you answered only 1 day ago.

**Comment by franzheidl** [comment]
Just released 0.0.5, hope this works for everyone! 👍
